libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -767,10 +767,52 @@ void StackLayoutGenerator::fillInJunk(CFG::BasicBlock const& _block)
 		createStackLayout(_source, _target, swap, dupOrPush, pop);
 		return opGas;
 	};
+	/// @returns the number of junk slots to be prepended to @a _targetLayout for an optimal transition from
+	/// @a _entryLayout to @a _targetLayout.
+	auto getBestNumJunk = [&](Stack const& _entryLayout, Stack const& _targetLayout) -> size_t {
+		size_t bestCost = evaluateTransform(_entryLayout, _targetLayout);
+		size_t bestNumJunk = 0;
+		size_t maxJunk = _entryLayout.size();
+		for (size_t numJunk = 1; numJunk <= maxJunk; ++numJunk)
+		{
+			size_t cost = evaluateTransform(_entryLayout, Stack{numJunk, JunkSlot{}} + _targetLayout);
+			if (cost < bestCost)
+			{
+				bestCost = cost;
+				bestNumJunk = numJunk;
+			}
+		}
+		return bestNumJunk;
+	};
+
+	if (_functionInfo && !_functionInfo->canContinue && _block.allowsJunk())
+	{
+		size_t bestNumJunk = getBestNumJunk(
+			_functionInfo->parameters | ranges::views::reverse | ranges::to<Stack>,
+			m_layout.blockInfos.at(&_block).entryLayout
+		);
+		if (bestNumJunk > 0)
+			addJunkRecursive(&_block, bestNumJunk);
+	}
+
 	/// Traverses the CFG and at each block that allows junk, i.e. that is a cut-vertex that never leads to a function
 	/// return, checks if adding junk reduces the shuffling cost upon entering and if so recursively adds junk
 	/// to the spanned subgraph.
 	util::BreadthFirstSearch<CFG::BasicBlock const*>{{&_block}}.run([&](CFG::BasicBlock const* _block, auto _addChild) {
+		if (_block->allowsJunk())
+		{
+			auto& blockInfo = m_layout.blockInfos.at(_block);
+			Stack entryLayout = blockInfo.entryLayout;
+			size_t bestNumJunk = getBestNumJunk(
+				entryLayout,
+				_block->operations.empty() ? blockInfo.exitLayout : m_layout.operationEntryLayout.at(&_block->operations.front())
+			);
+			if (bestNumJunk > 0)
+			{
+				addJunkRecursive(_block, bestNumJunk);
+				blockInfo.entryLayout = entryLayout;
+			}
+		}
```

Comment:
This is basically just refactoring - I can't think of a case that would be affected by this change.
It's just a matter of whether to consider adding junk to the current block, when entering it, or to the exit just before exiting, the result will be the same. But doing it when entering will make things easier, once we for example add a switch exit for implementing jump tables.

That being said, this will now be run even when the jump to the current block is unconditional, in which case the entry layout should always be identical to the next required layout - I'll add a shortcut condition, if the layouts are already identical, in that case adding no junk will always be optimal anyways.